### ISSUE TITLE
Post-processing responses directly in `@endpoint` methods

### ISSUE THREAD

**Issue by DefiDebauchery** [pmateusz/meatie#69]
The section in the README about postprocessing HTTP responses admittedly didn't make much sense to me, but it did point to the Option modules, which I thought I'd try utilizing.

In my client definition, there's an API endpoint that returns JSON that, if possible, would prefer to just return one value from. 

```py
class MyClient(Client):
    async def extract_nonce(self, response: dict) -> str:
        return response['nonce']

    @endpoint("/api/nonce", body(json=extract_nonce), method="POST")
    async def get_nonce(self) -> list:
        ...
```

I'm clearly not advanced-enough in python to really understand what this error is telling me it wants, but it reads
```py
Expected type '(Any) -> dict[str, Any] | (Any) -> Awaitable[dict[str, Any]] | None', got '(self: MyClient, response: dict) -> Coroutine[Any, Any, str]' instead 
```
I didn't see any examples of using the `body` Option, so I'm not entirely sure what format it's looking for here.

**Edit**: Although after re-reading `descriptor.py`, it seems this is a function for decoding the response itself, not so much what to do after doing so (though theoretically it should be able to...?)

Either way, am I on the right track at all, or would I be unable to transform responses in this way?

**Comment by pmateusz** [comment]
Apologies if you got confused by reading the documentation of the options module. It was not my intention.

The type hint error message you posted explains the `extract_nonce`function passed `body(json=extract_nonce)` is expected to return `dict[str, Any]` or its awaitable. The type hint should be improved in the meatie library, because a valid JSON is also a string or an array.

The primary use `body(json=...)` is to adjust JSON deserialization or handle situations where API sends corrupted JSONs (for instance, missing `}` or `]`) that we still need to process. 

Below I posted suggested implementation for the issue you posted.

```
class MyClient(Client):
    async def get_nonce(self) -> str:
	response = await self._get_nonce()
        return response['nonce']

    @endpoint("/api/nonce", method="POST")
    async def _get_nonce(self) -> list:
        ...
```

**Comment by DefiDebauchery** [comment]
@pmateusz I appreciate the clarification; I see that using `body` is not helpful for what I am looking to do.

However, Is it a rigid requirement for `meatie` that each `@endpoint` method to do nothing but serve as a stub for the direct response type? There are so many situations in which the actual API call is only a small part of the logic -- for instance, a `DELETE` request whose success may only come in the form of an HTTP status code (which downstream code shouldn't need to care directly about); having to make two methods for each endpoint in this way would lead to classes that are unnecessarily large.

```py
    @endpoint('/v1/multisig-transactions/{tx_hash}', method='DELETE')
    async def delete_multisig_transaction(self, tx_hash: str, response: AsyncStatus) -> bool:
        # Somehow get AsyncStatus into this method

        if response.status == 204:
            return True

        if response.status == 404:
            return False

        if response.status == 400:
            err = await status.json()
            raise ValueError(err)
```

Unfortunately my Python leans more rudimentary than how this helpful library is set up, otherwise I'd try to make any direct suggestions, but if there's a starting point you could provide, I'd be happy to give it a crack if needed.

**Comment by pmateusz** [comment]
Methods decorated with `@endpoint` can return `AsyncResponse`. It is a protocol defined in the library that represents an HTTP response. `AsyncResponse` protocol includes standard methods for accessing the response body, either bytes or text, and the HTTP status code.  The implementation of the protocol depends on the HTTP client library being used.

I am sharing an example that implements the use case you described:

```
import pytest
from meatie import AsyncResponse, HttpStatusError, endpoint


class Client:
 @endpoint("/v1/multisig-transactions/{tx_hash}")
 async def delete_multisig_transaction(self, tx_hash: str) -> AsyncResponse:
 ...
```

Full source: https://github.com/pmateusz/meatie/blob/master/tests/examples/aiohttp/test_wrap_client.py

I discovered a bug in handling `AsyncResponse` while answering this question. The bug is fixed in version v0.1.13. Before using the code sample, please install the latest version of the library.
 
`@endpoint` is technically a class descriptor (it is built using analogous mechanisms as @classmethod). Its purpose is to execute an HTTP request described by the method signature (parameters and the return type).

The idea of having methods empty is not a coincidence. The library makes it difficult for developers to combine  communication with REST API (1), parsing responses (2), and processing responses (3) within the same class (or layer). Essentially, doing all three things together in my opinion makes testing code more complex than it needs to be. Instead, Meatie takes care of (1) and (2) using empty methods. There is no need for unit testing that part. Hopefully, testing code for processing responses (3) is also easier because it can be done in isolation and Meatie clients are straightforward to mock.

**Comment by DefiDebauchery** [comment]
> The idea of having methods empty is not a coincidence. The library makes it difficult for developers to combine communication with REST API (1), parsing responses (2), and processing responses (3) within the same class (or layer).

Okay, so you would say that `meatie` explicitly does not intend to allow the option for these to be combined.

I can definitely see situations where that is preferable (implementing meatie into one's own packaged API, for instance). My (and I would imagine many people's) use is a little more flexible in that regard where we just want to create quick client APIs, and adding a separate layer on top would make things more complex than allowing inline manipulation.

In any case, while I was able to modify the `endpoint` method to actually enter the method (by returning a wrapped method that directly calls the original function with the response arg), it would clearly require several changes across the library since the return type of the decorated method is central to the response handling.

To be clear, I would still want this and believe it would make `meatie` a de facto library for standing up API clients incredibly easily, but given your comments and having dug into what it would require to make this happen, I'll settle for the "double-method" strategy originally proposed.
